### PR TITLE
feat: implement spanner::Instance class.

### LINKED ISSUE #565: Implement an Instance class analogous to spanner::Database
In the admin interfaces we need to refer to `Instances` with names like `"projects/my-project/instances/my-instance"`.  This suggests we should create a `spanner::Instance` class analogous to `spanner::Database`.

### PR DESCRIPTION
**Pull request by coryan**
This class represents a Cloud Spanner instance name, just like
`spanner::Database` represents a Cloud Spanner database name.

Fixes #565.

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp-spanner/577)

### CODE REVIEW

File: google/cloud/spanner/instance.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+
+/**
+ * This class identifies a Cloud Spanner Instance
+ *
+ * A Cloud Spanner instance is identified by its `project_id` and `instance_id`.
```
Comment (devjgm) [inline google/cloud/spanner/instance.h]: Oh, this is good. Now it makes me think that the `Database` class should change to pair an `Instance` + a `database_id`. And `Database::ParentName` should change to an `Instance` getter. None of these changes are needed in this PR, but if you like this idea, we could file an issue to "fix" `Database` to take advantage of this new class more.

File: google/cloud/spanner/instance.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+   * Returns the fully qualified instance name as a string of the form:
+   * "projects/<project-id>/instances/<instance-id>"
+   */
+  std::string FullName() const;
+
+  /**
+   * Returns the fully qualified name of the instance's parent of the form:
+   * "projects/<project-id>"
+   */
+  std::string ParentName() const;
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/spanner/instance.h]: Unclear, and now I suspect it won't be needed. I removed it, easy to add back if we actually need it.

File: google/cloud/spanner/instance.h (hunk 1 above)
Comment (coryan) [reply google/cloud/spanner/instance.h]: Filed #586

**Comment by coryan** [comment]
Rebased to resolve conflicts. PTAL.

**Pull request merged**